libzerocash/Coin.cpp

```diff
@@ -22,6 +22,48 @@ Coin::Coin(): addr_pk(), cm(), rho(rho_size), r(zc_r_size), coinValue(v_size) {
 
 }
 
+Coin::Coin(const std::string bucket, Address& addr): addr_pk(), cm(), rho(rho_size), r(zc_r_size), coinValue(v_size) {
+    // Retreive and decode the private key
+    ECIES<ECP>::PrivateKey decodedPrivateKey;
+    decodedPrivateKey.Load(StringStore(addr.getEncryptionSecretKey()).Ref());
+
+    // Create the decryption session
+    AutoSeededRandomPool prng;
+    ECIES<ECP>::Decryptor decrypt(decodedPrivateKey);
+
+    // Convert the input string into a vector of bytes
+    std::vector<byte> bucket_bytes(bucket.begin(), bucket.end());
+
+    // Construct a temporary object to store the plaintext, large enough
+    // to store the plaintext if it were extended beyond the real size.
+    std::vector<unsigned char> plaintext;
+    // Size as needed, filling with zeros.
+    plaintext.resize(decrypt.MaxPlaintextLength(decrypt.CiphertextLength(v_size + zc_r_size + rho_size)), 0);
```

Comment:
> It sounds like if you encrypt and then decrypt, the output cleartext may be larger than the input cleartext

Not really. If someone encrypts a *different* plaintext than what we expect, that's what I'm concerned about. There are effectively three sizes to be concerned about here: our desired plaintext length, the supplied ciphertext length, and the decrypted plaintext length. Ideally the desired and decrypted are the same, but out of caution we use these methods to give us buffer room to make sure we're not getting rused.